File: metalios/listios/StorageStarted/pan3d/frame3d/dis/FrameBuildSprite.m
//
//  FrameBuildSprite.m
//  iosgl
//
//  Created by pan3dme on 2020/11/16.
//  Copyright © 2020 zhao. All rights reserved.
//
#import "FrameBuildSprite.h"
#import "Scene3D.h"
#import "FrameBuildShader.h"
#import "ProgrmaManager.h"
#import "Context3D.h"
@implementation FrameBuildSprite

- (instancetype)init:(Scene3D *)val
{
    self =[super init:val];
    [self inidShader];
    return self;
}
-(void)inidShader
{
    [self.scene3D.progrmaManager registe:FrameBuildShader.shaderStr shader3d: [[FrameBuildShader alloc]init:self.scene3D]];
    self.shader3d=  [self.scene3D.progrmaManager getProgram:FrameBuildShader.shaderStr];
}

-(void)setFrameNodeUrl:(FrameNodeVo*)nodeVo;
{
    [super setFrameNodeUrl:nodeVo];
    FrameBuildSprite* this=self;
    this.groupItem =[[NSMutableArray alloc] init];
    Display3DSprite* dis =[[Display3DSprite alloc]init:this.scene3D];
    [dis setObjUrl:nodeVo.resurl];
    NSDictionary* info =nodeVo.materialInfoArr[0];
    [dis setPicUrl:  [info valueForKey:@"url"]];
    [this.groupItem addObject:dis];
    
}
- (void)upFrame
{
    FrameBuildSprite* this=self;
    if( this.sceneVisible){
        for (int i=0;i<this.groupItem.count;i++){
            [self drawTempDisplay:this.groupItem[i]];
        }
    }
}

-(void)drawTempDisplay:(Display3DSprite*)display
{
    
    if(display.objData==nil){
        return;
    }
    self.objData= display.objData;
 
    
    
    
    id<MTLRenderCommandEncoder> renderEncoder=self.scene3D.context3D.renderEncoder;
    [self.shader3d mtlSetProgramShader];
    [self setVc];
    
    [renderEncoder setVertexBuffer: self.objData.mtkvertices
                            offset:0
                           atIndex:0];
    
    [renderEncoder setVertexBuffer: self.objData.mtkuvs
                            offset:0
                           atIndex:1];
    
    
      [renderEncoder setFragmentTexture:display.textureRes.mtlTexture
                                atIndex:0];
  
    [renderEncoder drawIndexedPrimitives:MTLPrimitiveTypeTriangle
                              indexCount: self.objData.mtkindexCount
                               indexType:MTLIndexTypeUInt32
                             indexBuffer: self.objData.mtkindexs
                       indexBufferOffset:0];
    
}
- (void)setVc;
{
 
    id<MTLRenderCommandEncoder> renderEncoder=self.scene3D.context3D.renderEncoder;
    [self.scene3D.context3D setMatrixVc:self.scene3D.camera3D.modelMatrix renderEncoder:renderEncoder idx:2];
    [self.scene3D.context3D setMatrixVc: self.posMatrix3d renderEncoder:renderEncoder idx:3];
}
@end

File: android/android/app/src/main/java/com/pan3d/frame3d/dis/FrameBuildSprite.java
package com.pan3d.frame3d.dis;

import android.util.Log;

import com.pan3d.base.ObjData;
import com.pan3d.core.Context3D;
import com.pan3d.display.Display3DSprite;
import com.pan3d.frame3d.FrameNodeVo;
import com.pan3d.program.Shader3D;
import com.pan3d.scene.Scene3D;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

public class FrameBuildSprite  extends FrameBaseDisplay {
    private static final String TAG ="FrameBuildSprite" ;
    public FrameBuildSprite(Scene3D val) {
        super(val);
         scene3D.progrmaManager.registe(FrameBuildShader.FrameBuildShader,new FrameBuildShader(scene3D));
        this.shader3D= scene3D.progrmaManager.getProgram(FrameBuildShader.FrameBuildShader);
    }
    @Override
    public void setFrameNodeUrl(FrameNodeVo $vo) {
        super.setFrameNodeUrl($vo);
        this.groupItem = new ArrayList<>();
        Display3DSprite $dis = new Display3DSprite( scene3D);
        $dis.setObjUrl($vo.resurl);
        try {
            JSONObject objInfo=(JSONObject)$vo.materialInfoArr.get(0);
            $dis.setPicUrl(objInfo.getString("url"));
        } catch (JSONException e) {

        }

        this.groupItem.add($dis);
    }
    @Override
    public void upData() {
        super.upData();
       if( this.sceneVisible){
           for (int i=0;i<this.groupItem.size();i++){
               this.drawTempDisplay(this.groupItem.get(i));
           }
       }

    }
    private void drawTempDisplay(Display3DSprite display) {
        if(display.objData==null){
            return;
        }
        ObjData objData= display.objData;

        Log.d(TAG, "drawTempDisplay: ");
        Context3D ctx= scene3D.context3D;
        ctx.setProgame(this.shader3D.program);
        this.setVc();
        ctx.setVa(this.shader3D,"pos",3,objData.vertexBuffer);
        ctx.setVa(this.shader3D,"v2Uv",2,objData.uvBuffer);
        ctx.setRenderTexture(this.shader3D,"fs0",display.baseTextureRes.textTureInt,0);
        ctx.drawCall(display.objData.indexBuffer,objData.treNum);

    }

    protected void setVc()
    {
        Context3D ctx= scene3D.context3D;
        ctx.setVcMatrix4fv(this.shader3D, Shader3D.vpMatrix3D, scene3D.camera3D.modelMatrix.m);
        ctx.setVcMatrix4fv(this.shader3D,"posMatrix3D",this.posMatrix3d.m);

    }


}
